Sources/NIO/Selector.swift

```diff
@@ -372,12 +372,14 @@ final class Selector<R: Registration> {
                 // We are not interested in the result
                 _ = Glibc.read(timerfd, &val, MemoryLayout<UInt>.size)
             default:
-                let registration = registrations[Int(ev.data.fd)]!
-                try body(
-                    SelectorEvent(
-                        readable: (ev.events & Epoll.EPOLLIN.rawValue) != 0 || (ev.events & Epoll.EPOLLERR.rawValue) != 0 || (ev.events & Epoll.EPOLLRDHUP.rawValue) != 0,
-                        writable: (ev.events & Epoll.EPOLLOUT.rawValue) != 0 || (ev.events & Epoll.EPOLLERR.rawValue) != 0 || (ev.events & Epoll.EPOLLRDHUP.rawValue) != 0,
-                        registration: registration))
+                // If the registration is not in the Map anymore we deregistered it during the processing of whenReady(...). In this case just skip it.
+                if let registration = registrations[Int(ev.data.fd)] {
+                    try body(
+                        SelectorEvent(
+                            readable: (ev.events & Epoll.EPOLLIN.rawValue) != 0 || (ev.events & Epoll.EPOLLERR.rawValue) != 0 || (ev.events & Epoll.EPOLLRDHUP.rawValue) != 0,
+                            writable: (ev.events & Epoll.EPOLLOUT.rawValue) != 0 || (ev.events & Epoll.EPOLLERR.rawValue) != 0 || (ev.events & Epoll.EPOLLRDHUP.rawValue) != 0,
+                            registration: registration))
+                }
```

Comment:
@Lukasa @weissi I wonder if we also want to call `epoll_ctl_del` when the if let fails to be sure we never get a wake-up again. Same is true for the code for KQueue. WDYT ?

Reply:
I don't think that'll do anything.

The issue here specifically is that `events` is a collection that does not go via the kernel to get the next element. This means that if `body` ends up calling out to code that calls `unregister`, we'll update `registrations`, but not make any changes to `events`.  Setting `EPOLL_CTL_DEL` for that FD is not helpful, because we *already* called it in `deregister`: it's just that it hasn't had any effect yet.

Reply:
Thats true... that said we could call `epoll_ctl` with `EPOLL_CTL_DEL` and verify that it returns `ENOENT` when this happens (when in debug mode). This way we at least ensure this is the problem here. WDYT ?

Reply:
I'm ok doing it in debug mode, but only in debug mode.

Reply:
@Lukasa ok let me do this... better crashing hard in debug mode 👍